cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -107,7 +110,16 @@ class TypeInferrer {
       } else if (is_floating(type->id())) {
         ++float_count_;
       } else if (type->id() == Type::TIMESTAMP) {
-        ++timestamp_count_;
+        auto type2 = std::dynamic_pointer_cast<TimestampType>(type);
+        if (type2->unit() == TimeUnit::NANO) {
+          ++timestamp_nano_count_;
+        } else if (type2->unit() == TimeUnit::MICRO) {
+          ++timestamp_micro_count_;
+        } else if (type2->unit() == TimeUnit::MILLI) {
+          ++timestamp_milli_count_;
+        } else if (type2->unit() == TimeUnit::SECOND) {
+          ++timestamp_second_count_;
+        }
```

Comment:
If the scalar type is some other unit (or unitless? can't remember how we're handling that at the moment) then you should raise in an `else` branch